File: prohance.py
from time import sleep
import datetime
import re
import json
import redis
from rq import use_connection
from configparser import ConfigParser
import shlex, subprocess

def check_time():
    current_time = datetime.datetime.now()
    current_time = str(current_time)[11:]
    time_elements = []
    time_elements = current_time.split(':')
    if int(time_elements[0]) >= 25:
        write_log("Its 08:00 PM hence killing the script")
        write_log("#SCRIPT END")
        return True

def get_manager_id(mana_list):
    cfg = ConfigParser()
    cfg.read('/home/dhinesh/Credentials/Prohance/managerListCfg.ini')
    manager = mana_list.pop()
    write_log("Manger assined for this task is "+manager)
    mana_list.insert(0, manager)
    mana_id = cfg.get('selector', manager)
    return mana_id, mana_list

def write_log(msg):
    date = datetime.datetime.now()
    date = str(date)[:10]
    date = re.sub(r'-', '_', date, re.I|re.S)
    filename = '/home/dhinesh/Prohance/Logs/prohance_logs_' + date + '.log'
    msg = 'KFetcher:' + str(datetime.datetime.now()) + ': ' + msg + "\n"
    with open(filename, "a") as f:
        f.write(msg)



index = ''
non_bank_list = ['Muzammil', 'Bindu', 'Chandrakala', 'Vanita']
bank_list = ['Shankar', 'Harishkumar', 'Varun', 'Peter', 'Premkumar', 'Sahana', 'Shilpa', 'Shivanna', 'Shobha', 'Shreedevi', 'Sreedhara']
while(True) :
    my_connection = redis.StrictRedis(host='localhost', port=6379, db=0, password=None)
    if (check_time()) :
        break
    result = my_connection.lpop('test_qu')
    if (result == None):
        write_log("Queue is empty. Sleeping")
        sleep(5)
    else:
        try:
            res_obj = re.search(r'^b\'(.*?)\'$', str(result), re.I)
            result = res_obj.group(1)
            write_log("JSON found for the task - "+str(result))
            input_data = json.loads(result)
            if input_data['processingType'] == 'STATEMENT':
                manager_id, bank_list = get_manager_id(bank_list)
            elif input_data['processingType'] == 'Financial Statement' or input_data['processingType'] == 'ITRV Statement':
                manager_id, non_bank_list = get_manager_id(non_bank_list)
            txt = open('/home/dhinesh/Credentials/Prohance/serial')
            index = txt.read()
            index = int(index) + 1
            write_log("Index for this task - "+str(index))
            try:
                arguments = str(result) + "\n" + str(manager_id) + "\n" + str(index)
                write_log("#START_TASK_CREATION")
                p = subprocess.Popen(['perl', 'task.pl', arguments], stdout=subprocess.PIPE)
            except Exception as e:
                print ("Exception occurred :"+str(e))
        except Exception as e:
            print ("Exception occurred : "+str(e))

File: Prohance.pl
#!/usr/bin/perl
use strict;
use Redis::Client;
use Redis;
use JSON;
use Config::Simple;
use TaskCreator qw(create_task write_log);

#my $conn = Redis::Client->new( host => 'localhost', port => 6379 );
my $b = '{"processingType" : "STATEMENT", "perfiosTransactionId" : "A123", "downloadUrl" : "https://www.perfios.com/KuberaVault/...", "environment" : "PRODUCTION"}';
#my $len = $conn->rpush(my_list => $b);
#print "$len\n";
my $r = Redis->new(server => 'localhost:6379');
#$r->auth('falcon123');
#$r->lpush('my_list', $b);
my $er = $r->lpop('test_qu');
print $er . "\n";


=h
my ($manager_id, $serial, $file);
my @bank_list = ('Shankar', 'Harishkumar', 'Varun', 'Peter', 'Premkumar', 'Sahana', 'Shilpa', 'Shivanna', 'Shobha', 'Shreedevi', 'Sreedhara');
my @non_bank_list = ('Muzammil', 'Bindu', 'Chandrakala', 'Vanita');
my $flag = 1;
while ($flag) {
    my $client = Redis::Client->new(host => 'localhost', port => 6379);
    if (check_time()) {
        $flag = 0;
        next;
    }
    my $result = $client->rpop('test_qu');
    if (!$result) {
        write_log("Queue is empty. Sleeping for 5 secs.");
        sleep(5);
    } else {
        write_log("Data recieved - $result");
        my $input_data = decode_json($result);
        if ($input_data->{processingType} eq 'STATEMENT') {
            $manager_id = get_manager_id('bank');
        } elsif ($input_data->{processingType} eq 'Financial Statement' or $input_data->{processingType} eq 'ITR Statement') {
            $manager_id = get_manager_id('non_bank');
        }
        $file = '/home/dhinesh/Credentials/Prohance/serial';
        open FH, $file;
        $serial .= $_ while(<FH>);
        close FH;
        $serial = int($serial) + 1;
        $file = '/home/dhinesh/Credentials/Prohance/serial';
        open F1, ">" . $file;
        binmode F1, ':utf8';
        print F1 $serial;
        close F1;
        write_log("Serial number for this task - $serial");
        write_log("#CREATING_TASK");
        create_task($input_data, $manager_id, $serial);
        write_log("#TASK_CREATION_DONE");
        $serial = '';
    }
}

sub check_time {
    my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = localtime();
    if ($hour >= 20) {
        print "Its 08.00PM Hence killing the script";
        return 1;
    } else {
        return 0;
    }
}

sub get_manager_id {
    my $type = shift;
    my $manager = '';
    my $cfg = new Config::Simple('/home/dhinesh/Credentials/Prohance/managerListCfg.ini');
    if ($type eq 'bank') {
        $manager = shift(@bank_list);
        push (@bank_list, $manager);
    } elsif ($type eq 'non_bank') {
        $manager = shift(@non_bank_list);
        push (@non_bank_list, $manager);
    }
    write_log("Manager assigned for this task - $manager");
    my $id = $cfg->param("selector.$manager");
    return $id;
}
=cut
